File: leetcode/301-400/307-range-sum-query-mutable/bench/fenwick.py
# Benchmark mirror — LeetCode 307, Range Sum Query (Mutable).
# Same Fenwick tree, same LCG-generated operation script, same masked sink as
# fenwick.kara. See ../README.md § Benchmarks.
#
# The sink masks rather than taking a modulus: Python's `%` FLOORS while
# C/Rust/Go/kara truncate toward zero, so a modulo sink over a signed running
# total prints a different number here than in every other mirror (measured on
# #303). Masking is two's-complement in all five languages.


def main():
    n = 65536
    ops = 200000
    passes = 110

    tree = [0] * (n + 1)
    data = [0] * n
    kind = [0] * ops
    opa = [0] * ops
    opb = [0] * ops

    state = 20307
    for k in range(ops):
        state = (state * 1103515245 + 12345) % 2147483648
        t = state % 2
        state = (state * 1103515245 + 12345) % 2147483648
        x = state % n
        state = (state * 1103515245 + 12345) % 2147483648
        y = state % n
        kind[k] = t
        if t == 0:
            opa[k], opb[k] = x, y % 2001 - 1000
        elif x <= y:
            opa[k], opb[k] = x, y
        else:
            opa[k], opb[k] = y, x

    checksum = 0
    for _ in range(passes):
        for k in range(ops):
            if kind[k] == 0:
                i = opa[k]
                delta = opb[k] - data[i]
                data[i] = opb[k]
                x = i + 1
                while x <= n:
                    tree[x] += delta
                    x += x & -x
            else:
                total = 0
                hi = opb[k] + 1
                while hi > 0:
                    total += tree[hi]
                    hi -= hi & -hi
                lo = opa[k]
                while lo > 0:
                    total -= tree[lo]
                    lo -= lo & -lo
                checksum = (checksum + total) & 0x3FFFFFFF
    print(f"checksum {checksum}")


main()

File: leetcode/301-400/307-range-sum-query-mutable/bench/fenwick.rs
// Benchmark mirror — LeetCode 307, Range Sum Query (Mutable).
// Same Fenwick tree, same LCG-generated operation script, same masked sink as
// fenwick.kara. See ../README.md § Benchmarks.

fn main() {
    let n: i64 = 65536;
    let ops: i64 = 200000;
    let passes: i64 = 110;

    let mut tree: Vec<i64> = vec![0; (n + 1) as usize];
    let mut data: Vec<i64> = vec![0; n as usize];
    let mut kind: Vec<i64> = Vec::with_capacity(ops as usize);
    let mut opa: Vec<i64> = Vec::with_capacity(ops as usize);
    let mut opb: Vec<i64> = Vec::with_capacity(ops as usize);

    let mut state: i64 = 20307;
    for _ in 0..ops {
        state = (state * 1103515245 + 12345) % 2147483648;
        let t = state % 2;
        state = (state * 1103515245 + 12345) % 2147483648;
        let x = state % n;
        state = (state * 1103515245 + 12345) % 2147483648;
        let y = state % n;
        kind.push(t);
        if t == 0 {
            opa.push(x);
            opb.push(y % 2001 - 1000);
        } else if x <= y {
            opa.push(x);
            opb.push(y);
        } else {
            opa.push(y);
            opb.push(x);
        }
    }

    let mut checksum: i64 = 0;
    for _ in 0..passes {
        for k in 0..ops as usize {
            if kind[k] == 0 {
                let i = opa[k];
                let delta = opb[k] - data[i as usize];
                data[i as usize] = opb[k];
                let mut x = i + 1;
                while x <= n {
                    tree[x as usize] += delta;
                    x += x & -x;
                }
            } else {
                let mut total: i64 = 0;
                let mut hi = opb[k] + 1;
                while hi > 0 {
                    total += tree[hi as usize];
                    hi -= hi & -hi;
                }
                let mut lo = opa[k];
                while lo > 0 {
                    total -= tree[lo as usize];
                    lo -= lo & -lo;
                }
                checksum = (checksum + total) & 0x3FFFFFFF;
            }
        }
    }
    println!("checksum {}", checksum);
}
